File: autoids_-can-&-ethernet-intrusion-detection-system/src/simulation/central_processor.ts
import { CANPacket, EthernetPacket, Alert } from './types.js';
import { EthRingBuffer } from './eth_buffer.js';
import { gPTP } from './gptp.js';

export class CentralProcessor {
  private ethBuffer: EthRingBuffer;
  private alertCounter: number = 0;

  constructor(ethBuffer: EthRingBuffer) {
    this.ethBuffer = ethBuffer;
  }

  processAlert(canPacket: CANPacket): Alert {
    // Get context from Ethernet buffer (e.g., last 500ms)
    const contextStartTime = canPacket.timestamp - 500;
    const contextEndTime = canPacket.timestamp + 100;
    const contextPackets = this.ethBuffer.getItemsWithinTimeframe(contextStartTime, contextEndTime);

    // Simulate ML Classification
    let classification = "Unknown";
    let confidence = 0.0;

    const hasEthernetAttack = contextPackets.some(p => p.isAttack);

    if (hasEthernetAttack && canPacket.isAttack) {
      classification = `Ethernet -> CAN (${canPacket.id})`;
      confidence = 0.92 + Math.random() * 0.07;
    } else if (canPacket.isAttack) {
      classification = `Internal CAN (${canPacket.id})`;
      confidence = 0.85 + Math.random() * 0.1;
    } else {
      classification = "False Positive";
      confidence = 0.6 + Math.random() * 0.2;
    }

    this.alertCounter++;
    const currentTime = gPTP.getGlobalTimeMs();
    
    const alert: Alert = {
      id: `ALT-${this.alertCounter}-${currentTime}`,
      timestamp: currentTime,
      canPacket,
      ethernetContext: contextPackets,
      classification,
      confidence
    };

    return alert;
  }
}

File: REAL-IDS/cpp/src/central_processor.cpp
#include "real_ids/central_processor.hpp"

namespace real_ids {

CentralProcessor::CentralProcessor(EthRingBuffer* eth_buffer) : eth_buffer_(eth_buffer) {}

Alert CentralProcessor::process_alert(const CanPacket& can_packet, std::uint64_t processor_time_ms,
                                        bool production_mode) {
  const std::uint64_t context_start =
      can_packet.timestamp_ms > 500 ? (can_packet.timestamp_ms - 500) : 0;
  const std::uint64_t context_end = can_packet.timestamp_ms + 100;

  auto context = eth_buffer_->items_within_timeframe(context_start, context_end);

  bool has_eth_attack = false;
  for (const auto& p : context) {
    if (p.synthetic_attack_flag) {
      has_eth_attack = true;
      break;
    }
  }

  std::string classification = "Unknown";
  float confidence = 0.F;

  if (production_mode) {
    classification = has_eth_attack ? "CAN anomaly with suspicious ETH context"
                                    : "CAN timing anomaly (" + can_packet.id + ")";
    confidence = has_eth_attack ? 0.82F : 0.78F;
  } else if (has_eth_attack && can_packet.synthetic_attack_flag) {
    classification = "Ethernet -> CAN (" + can_packet.id + ")";
    confidence = 0.92F;
  } else if (can_packet.synthetic_attack_flag) {
    classification = "Internal CAN (" + can_packet.id + ")";
    confidence = 0.85F;
  } else {
    classification = "Anomaly (no attack metadata)";
    confidence = 0.75F;
  }

  ++alert_counter_;
  Alert alert;
  alert.id = "ALT-" + std::to_string(alert_counter_) + "-" + std::to_string(processor_time_ms);
  alert.timestamp_ms = processor_time_ms;
  alert.can_packet = can_packet;
  alert.ethernet_context = std::move(context);
  alert.classification = std::move(classification);
  alert.confidence = confidence;
  return alert;
}

}  // namespace real_ids
